### PR TITLE
commands/build: charm file name format support for bases (CRAFT-66)

### PR DESCRIPTION
**Pull request by cjp256**
Maintain current format as "legacy" until we have the (default) bases
configuration available in upcoming PRs. Note that only the legacy
path is utilized at runtime, with tests added for bases configuration
support.

Signed-off-by: Chris Patterson <chris.patterson@canonical.com>

### CODE REVIEW

File: charmcraft/commands/build.py [hunk 1]
```diff
@@ -53,6 +55,41 @@
...
+def format_charm_file_name(
+    *, charm_name: str, bases_config: Optional[BasesConfiguration] = None
+) -> str:
+    """Formulate charm file name.
+
+    :param charm_name: Name of charm.
+    :param bases_config: Bases configuration for charm.  None will use legacy
+        format that will be removed shortly.
+
+    :returns: File name string, including .zip extension.
```
Comment (sergiusens) [inline charmcraft/commands/build.py:84]: .zip?

**Review by sergiusens — APPROVED** [review]
Please ensure the docstring mentioned in the inline review is correct.

**Review by facundobatista — CHANGES_REQUESTED** [review]
Something to fix, a question, and a suggestion.

FTR, I will simplify the code. When you create a function that is only a line of code, used in only one place, and not even tested separatedly, why to have that line of code separated? It just adds overhead to the code reading.

File: charmcraft/commands/build.py [hunk 2]
```diff
@@ -53,6 +55,41 @@
...
+    )
+
+
+def _format_bases_config(bases_config: BasesConfiguration) -> str:
+    """Formulate charm string for bases configuration section."""
+    return "_".join([_format_run_on_base(r) for r in bases_config.run_on])
+
+
+def format_charm_file_name(
+    *, charm_name: str, bases_config: Optional[BasesConfiguration] = None
```
[the remark this answers is not in the document]
Comment (cjp256) [reply charmcraft/commands/build.py:76]: Both parameters will be required soon.

File: charmcraft/commands/build.py [hunk 3]
```diff
@@ -53,6 +55,41 @@
...
 
+def _format_architectures(architectures: List[str]) -> str:
+    """Formulate charm string for architectures list section."""
+    return "-".join(architectures)
+
+
+def _format_run_on_base(base: Base) -> str:
+    """Formulate charm string for base section."""
+    return "-".join(
+        [base.name, base.channel, _format_architectures(base.architectures)]
```
[the remark this answers is not in the document]
Comment (cjp256) [reply charmcraft/commands/build.py]: Done!

File: charmcraft/commands/build.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (cjp256) [reply charmcraft/commands/build.py]: Indeed, fixed!

Comment (cjp256) [reply charmcraft/commands/build.py]: Fixed!

File: charmcraft/commands/build.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cjp256) [reply charmcraft/commands/build.py]: Removed parameter naming.

**Review by cmatsuoka — APPROVED** [review]
Looks good!

**Review by facundobatista — APPROVED** [review]
Great, thanks!

**Pull request merged**